### DISCUSSION TITLE
api test returns bad request for upload pdf file

### DISCUSSION THREAD

**Discussion by darec89** [cypress-io/cypress#27281] [Questions and Help]
this is the code.

Please tell what is wrong.
Postman and swagger returns 200. 

// <reference types="cypress" />
function generateBoundary() {
  return `----WebKitFormBoundary${Math.random().toString(16).slice(2)}`;
}
function createMultipartBody(fileContent, boundary) {
  return `--${boundary}\r\n` +
    `Content-Disposition: form-data; name="user"\r\n\r\n` +
    `${JSON.stringify({ name: "aleks", surname: "markovic", email: "[aleks@gmail.com](mailto:aleks@gmail.com)" })}\r\n` +
    `--${boundary}\r\n` +
    `Content-Disposition: form-data; name="document_type"\r\n\r\n` +
    `authorization\r\n` +
    `--${boundary}\r\n` +
    `Content-Disposition: form-data; name="documents"; filename="5eda1f9076c29.pdf"\r\n` +
    `Content-Type: application/pdf\r\n\r\n` +
    `${fileContent}\r\n` +
    `--${boundary}--`;
}
describe('API POST endpoint with image parameter', function () {
  it('should upload an image successfully', () => {
    cy.fixture('5eda1f9076c29.pdf', 'base64').then((fileContent) => {
      const boundary = generateBoundary();
      cy.request({
        method: "POST",
        url: "http://localhost:8000/verify",
        headers: {
          accept: "application/json",
          "Content-Type": `multipart/form-data; boundary=${boundary}`,
        },
        body: createMultipartBody(fileContent, boundary)
      }).then((response) => {
        expect(response.status).to.equal(200);
      });
    });
  });
});

**Comment by portyu9** [comment]
Hi @darec89 

The problem is the PDF part of the multipart body.

`cy.fixture('5eda1f9076c29.pdf', 'base64')` gives you a **base64-encoded string**, but `createMultipartBody()` inserts that string directly into a part declared as:

```http
Content-Type: application/pdf
```

So the server receives base64 text where it expects the actual PDF bytes. Postman/Swagger are sending the real binary file.

I would avoid manually constructing the multipart boundary and body. Convert the fixture back to binary and use `FormData`, letting the browser generate the correct boundary:

```js
cy.fixture('5eda1f9076c29.pdf', 'base64').then((base64) => {
  const bytes = Cypress.Buffer.from(base64, 'base64')

  cy.window().then((win) => {
    const formData = new win.FormData()

    formData.append(
      'user',
      JSON.stringify({
        name: 'aleks',
        surname: 'markovic',
        email: 'aleks@gmail.com',
      }),
    )

    formData.append('document_type', 'authorization')

    const pdf = new win.Blob([bytes], {
      type: 'application/pdf',
    })

    formData.append('documents', pdf, '5eda1f9076c29.pdf')

    return win.fetch('http://localhost:8000/verify', {
      method: 'POST',
      headers: {
        Accept: 'application/json',
      },
      body: formData,
    })
  }).then((response) => {
    expect(response.status).to.eq(200)
  })
})
```

Do **not** manually set `Content-Type: multipart/form-data` here. `fetch()` will add it together with the boundary it generated.

One limitation is that browser `fetch()` is subject to CORS. If this endpoint is not accessible from the application's origin, use the same `FormData` approach in a Node-side `cy.task()` instead.

The full discussion was reviewed; there were no screenshots/attachments or replies to inspect.

Please mark this answer as accepted if it helped, thank you.
